src/V3DfgAstToDfg.cpp

```diff
@@ -0,0 +1,239 @@
+// -*- mode: C++; c-file-style: "cc-mode" -*-
+//*************************************************************************
+// DESCRIPTION: Verilator: Convert AstModule to DfgGraph
+//
+// Code available from: https://verilator.org
+//
+//*************************************************************************
+//
+// Copyright 2003-2022 by Wilson Snyder. This program is free software; you
+// can redistribute it and/or modify it under the terms of either the GNU
+// Lesser General Public License Version 3 or the Perl Artistic License
+// Version 2.0.
+// SPDX-License-Identifier: LGPL-3.0-only OR Artistic-2.0
+//
+//*************************************************************************
+// V3Dataflow's Transformations:
+//
+//*************************************************************************
+
+#include "config_build.h"
+#include "verilatedos.h"
+
+#include "V3Ast.h"
+#include "V3Error.h"
+#include "V3Global.h"
+
+#include "V3Dfg.h"
+#include "V3DfgPasses.h"
+
+namespace {
+
+// Create a DfgVertex out of a AstNodeMath. For most AstNodeMath subtypes, this can be done
+// automatically. For the few special cases, we provide specializations below
+template <typename Vertex, typename Node>  //
+Vertex* makeVertex(const Node* nodep, DfgGraph& dfg) {
+    static_assert(std::is_base_of<DfgVertex, Vertex>::value,
+                  "'Vertex' must be subclass of 'DfgVertex'");
+    static_assert(std::is_base_of<AstNodeMath, Node>::value,
+                  "'Node' must be subclass of 'AstNodeMath'");
+    return new Vertex{dfg, nodep->fileline(), static_cast<uint32_t>(nodep->width())};
+}
+
+// Currently unhandled nodes
+template <>  //
+DfgAtoN* makeVertex<DfgAtoN, AstAtoN>(const AstAtoN*, DfgGraph&) {
+    return nullptr;
+}
+
+template <>  //
+DfgCCast* makeVertex<DfgCCast, AstCCast>(const AstCCast*, DfgGraph&) {
+    return nullptr;
+}
+
+template <>  //
+DfgCompareNN* makeVertex<DfgCompareNN, AstCompareNN>(const AstCompareNN*, DfgGraph&) {
+    return nullptr;
+}
+
+template <>  //
+DfgSliceSel* makeVertex<DfgSliceSel, AstSliceSel>(const AstSliceSel*, DfgGraph&) {
+    return nullptr;
+}
+
+template <>  // TODO: Select into final dimension should work fine
+DfgArraySel* makeVertex<DfgArraySel, AstArraySel>(const AstArraySel*, DfgGraph&) {
+    return nullptr;
+}
```

Comment:
Anything that returns a nullptr here is left as AST and is not converted to DFG (unhandled), can you think of any more AstNodeMath classes that might be problematic and should be included here?

Reply:
Why are compares and casts problematic?   I'd think this should generally be anything imprure, and arrays as your code doesn't handle them.  What about function and member calls?

Reply:
I turned off casts because they change dtypes/widths in some non-trivial (for me) way, but they should not really exist at all at the point this is run, but being paranoid.

CompareNN operates on strings. In general should probably turn all this off on string (they don't really have a width), and float/double, I will do that.

Calls are not expressions (not AstNodeMath) so are currently covered by the default case here:
https://github.com/gezalore/verilator/blob/f76e701919462d2a1060742c12408d73f9880a93/src/V3DfgAstToDfg.cpp#L119-L123

I think I am missing the impure check, will add that.

Reply:
Yes, better to disable for a string dtype, as opposed string operators.

Reply:
I actually went the opposite direction and support only explicitly bit-vector like basic types and packed arrays of the same for now. On XiangShan I get the same results. We can add handling for more complex types later.